### PR TITLE
Use public names when printing dune files

### LINKED ISSUE #2425: default_implementation should enforce public_name
I used the internal name of a library for `default_implementation` rather than the `public_name`. I initially didn't notice because the library was vendored, but did unexpectedly notice when I pinned the package in `opam` and built that way, since the `dune-package` file then contains a name which isn't in scope.

If the library referenced has a `public_name`, shouldn't Dune either insist on its being used in the `default_implementation` field *or* should it silently convert it to the `public_name` when it writes a `dune-package` file?

### PR DESCRIPTION
**Pull request by rgrinberg**
Fix #2425 and the bug @NathanReb encountered with @avsm.

### CODE REVIEW

File: src/dune/lib.ml [hunk 1]
```diff
@@ -1755,15 +1756,36 @@ let to_dune_lib ({ name; info; _ } as lib) ~modules ~foreign_objects ~dir =
...
   let special_builtin_support = Lib_info.special_builtin_support info in
-  let default_implementation = Lib_info.default_implementation info in
   let known_implementations = Lib_info.known_implementations info in
   let foreign_archives = Lib_info.foreign_archives info in
+  let use_public_name ~lib_field ~info_field =
+    match
+      let open Option.O in
+      let* loc, _ = info_field in
+      let+ field = lib_field in
+      (loc, field)
```
Comment (diml) [inline src/dune/lib.ml]: That feels like a complicated version of `match info_field, lib_field with ...`

File: test/blackbox-tests/test-cases/vlib-default-impl/run.t [hunk 2]
```diff
@@ -33,29 +33,15 @@ Check that ambiguity is handled correctly.
 Check that default implementation data is installed in the dune package file.
   $ dune build --root dune-package
   Entering directory 'dune-package'
+  File "dune", line 4, characters 25-34:
+  4 |  (default_implementation a-default))
+                               ^^^^^^^^^
+  Error: Library "a-default" not found.
+  Hint: try: dune external-lib-deps --missing --root dune-package @@default
+  [1]
```
Comment (diml) [inline test/blackbox-tests/test-cases/vlib-default-impl/run.t]: This test should be updated, otherwise it's not longer testing what it was meant to test.

**Review by ghost — CHANGES_REQUESTED** [review]
Looks good apart from a test that needs updating

File: test/blackbox-tests/test-cases/vlib-default-impl/run.t (hunk 2 above)
Comment (rgrinberg) [reply test/blackbox-tests/test-cases/vlib-default-impl/run.t]: Done.

**Comment by rgrinberg** [comment]
Thanks for the review. I addressed your comments.

**Pull request merged**